File: commons-core/src/main/java/xyz/wagyourtail/commons/position/Pos2D.java
package xyz.wagyourtail.commons.position;

import java.util.Objects;

public class Pos2D {
    public static final Pos2D ZERO = new Pos2D(0, 0);

    public final double x;
    public final double y;

    public Pos2D(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public static Pos2D max(Pos2D p1, Pos2D p2) {
        return new Pos2D(Math.max(p1.x, p2.x), Math.max(p1.y, p2.y));
    }

    public static Pos2D min(Pos2D p1, Pos2D p2) {
        return new Pos2D(Math.min(p1.x, p2.x), Math.min(p1.y, p2.y));
    }

    public static Pos2D abs(Pos2D p) {
        return new Pos2D(Math.abs(p.x), Math.abs(p.y));
    }

    public Pos2D inverse() {
        return new Pos2D(-x, -y);
    }

    public Pos2D plus(Pos2D other) {
        return new Pos2D(x + other.x, y + other.y);
    }

    public Pos2D plus(Pos2 other) {
        return new Pos2D(x + other.x, y + other.y);
    }

    public Pos2D plus(double x, double y) {
        return new Pos2D(this.x + x, this.y + y);
    }

    public Pos2D plus(double scalar) {
        return new Pos2D(this.x + scalar, this.y + scalar);
    }

    public Pos2D minus(Pos2D other) {
        return new Pos2D(x - other.x, y - other.y);
    }

    public Pos2D minus(Pos2 other) {
        return new Pos2D(x - other.x, y - other.y);
    }

    public Pos2D minus(double x, double y) {
        return new Pos2D(this.x - x, this.y - y);
    }

    public Pos2D minus(double scalar) {
        return new Pos2D(this.x - scalar, this.y - scalar);
    }

    public Pos2D times(Pos2D other) {
        return new Pos2D(x * other.x, y * other.y);
    }

    public Pos2D times(Pos2 other) {
        return new Pos2D(x * other.x, y * other.y);
    }

    public Pos2D times(double x, double y) {
        return new Pos2D(this.x * x, this.y * y);
    }

    public Pos2D times(double scalar) {
        return new Pos2D(this.x * scalar, this.y * scalar);
    }

    public Pos2D div(Pos2D other) {
        return new Pos2D(x / other.x, y / other.y);
    }

    public Pos2D div(Pos2 other) {
        return new Pos2D(x / other.x, y / other.y);
    }

    public Pos2D div(double x, double y) {
        return new Pos2D(this.x / x, this.y / y);
    }

    public Pos2D div(double scalar) {
        return new Pos2D(x / scalar, y / scalar);
    }

    public Pos2D rem(Pos2D other) {
        return new Pos2D(x % other.x, y % other.y);
    }

    public Pos2D rem(double x, double y) {
        return new Pos2D(this.x % x, this.y % y);
    }

    public Pos2D rem(double scalar) {
        return new Pos2D(x % scalar, y % scalar);
    }

    public Pos2D mod(Pos2D other) {
        double x = this.x % other.x;
        double y = this.y % other.y;
        if (x < 0) x += other.x;
        if (y < 0) y += other.y;
        return new Pos2D(x, y);
    }

    public Pos2D mod(double x, double y) {
        double x2 = this.x % x;
        double y2 = this.y % y;
        if (x2 < 0) x2 += x;
        if (y2 < 0) y2 += y;
        return new Pos2D(x2, y2);
    }

    public Pos2D mod(double scalar) {
        double x = this.x % scalar;
        double y = this.y % scalar;
        if (x < 0) x += scalar;
        if (y < 0) y += scalar;
        return new Pos2D(x, y);
    }

    public Pos2 toInt() {
        return new Pos2((int) x, (int) y);
    }

    public Pos2 floor() {
        return new Pos2((int) Math.floor(x), (int) Math.floor(y));
    }

    public Pos2 ceil() {
        return new Pos2((int) Math.ceil(x), (int) Math.ceil(y));
    }

    public Pos2 round() {
        return new Pos2((int) Math.round(x), (int) Math.round(y));
    }

    public double magnitude() {
        return Math.sqrt(x * x + y * y);
    }

    public double magnitudeSquared() {
        return x * x + y * y;
    }

    public Pos2D normalize() {
        double mag = magnitude();
        return new Pos2D(x / mag, y / mag);
    }

    public double dot(Pos2D other) {
        return x * other.x + y * other.y;
    }

    @Override
    public String toString() {
        return "(" + x + ", " + y + ")";
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Pos2D)) return false;
        Pos2D pos2D = (Pos2D) o;
        return Double.compare(x, pos2D.x) == 0 && Double.compare(y, pos2D.y) == 0;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }

}

File: commons-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/position/Pos2D.kt
package xyz.wagyourtail.commonskt.position

import kotlin.math.*

open class Pos2D(val x: Double, val y: Double) {
    companion object {
        val ZERO = Pos2D(0.0, 0.0)
    }

    open operator fun unaryMinus(): Pos2D {
        return Pos2D(-x, -y)
    }

    operator fun plus(other: Pos2D): Pos2D {
        return Pos2D(x + other.x, y + other.y)
    }

    operator fun plus(other: Pos2): Pos2D {
        return Pos2D(x + other.x, y + other.y)
    }

    fun plus(x: Double, y: Double): Pos2D {
        return Pos2D(this.x + x, this.y + y)
    }

    open operator fun plus(scalar: Double): Pos2D {
        return Pos2D(this.x + scalar, this.y + scalar)
    }

    operator fun minus(other: Pos2D): Pos2D {
        return Pos2D(x - other.x, y - other.y)
    }

    operator fun minus(other: Pos2): Pos2D {
        return Pos2D(x - other.x, y - other.y)
    }

    fun minus(x: Double, y: Double): Pos2D {
        return Pos2D(this.x - x, this.y - y)
    }

    open operator fun minus(scalar: Double): Pos2D {
        return Pos2D(this.x - scalar, this.y - scalar)
    }

    operator fun times(other: Pos2D): Pos2D {
        return Pos2D(x * other.x, y * other.y)
    }

    operator fun times(other: Pos2): Pos2D {
        return Pos2D(x * other.x, y * other.y)
    }

    fun times(x: Double, y: Double): Pos2D {
        return Pos2D(this.x * x, this.y * y)
    }

    open operator fun times(scalar: Double): Pos2D {
        return Pos2D(this.x * scalar, this.y * scalar)
    }

    operator fun div(other: Pos2D): Pos2D {
        return Pos2D(x / other.x, y / other.y)
    }

    operator fun div(other: Pos2): Pos2D {
        return Pos2D(x / other.x, y / other.y)
    }

    fun div(x: Double, y: Double): Pos2D {
        return Pos2D(this.x / x, this.y / y)
    }

    open operator fun div(scalar: Double): Pos2D {
        return Pos2D(x / scalar, y / scalar)
    }

    open fun toInt(): Pos2 {
        return Pos2(x.toInt(), y.toInt())
    }

    open fun floor(): Pos2 {
        return Pos2(floor(x).toInt(), floor(y).toInt())
    }

    open fun ceil(): Pos2 {
        return Pos2(ceil(x).toInt(), ceil(y).toInt())
    }

    open fun round(): Pos2 {
        return Pos2(round(x).toInt(),round(y).toInt())
    }

    open val magnitude: Double
        get() {
            return sqrt(magnitudeSquared)
        }

    open val magnitudeSquared: Double
        get() {
            return x * x + y * y
        }

    val normalized: Pos2D
        get() {
            val mag = magnitude
            return Pos2D(x / mag, y / mag)
        }

    fun dot(other: Pos2D): Double {
        return x * other.x + y * other.y
    }

    override fun toString(): String {
        return "($x, $y)"
    }

    override fun equals(other: Any?): Boolean {
        return other is Pos2D && other.x == x && other.y == y
    }

    override fun hashCode(): Int {
        var result = x.hashCode()
        result = 31 * result + y.hashCode()
        return result
    }

}
